contracts/pcv/curve/CurveSwapper.sol

```diff
@@ -0,0 +1,195 @@
+                        address coins3 = ICurvePool(_basePool).coins(3);
+                        require(_tokenReceived == coins3, "CurveSwapper: tokenReceived not found");
+                        _j = 4;
+                    }
+                }
+            }
+        } else {
+            require(_tokenSpent == coins1, "CurveSwapper: tokenSpent not found");
+            _basePool = coins0;
+            if (coins0 == ERC20_3CRV) _basePool = CURVE_3POOL;
+
+            _j = 0;
+            bool i1 = ICurvePool(_basePool).coins(0) == _tokenSpent;
+            if (i1) _i = 1;
+            else {
+                bool i2 = ICurvePool(_basePool).coins(1) == _tokenSpent;
+                if (i2) _i = 2;
+                else {
+                    bool i3 = ICurvePool(_basePool).coins(2) == _tokenSpent;
+                    if (i3) {
+                        _i = 3;
+                    } else {
+                        address coins3 = ICurvePool(_basePool).coins(3);
+                        require(_tokenSpent == coins3, "CurveSwapper: tokenSpent not found");
+                        _i = 4;
+                    }
+                }
+            }
+        }
+
+        i = _i;
+        j = _j;
+    }
+
+    /// @notice performs a swap
+    function swap()
+        external
+        override
+        whenNotPaused
+        hasAnyOfFourRoles(
+            TribeRoles.GUARDIAN,
+            TribeRoles.SWAP_ADMIN_ROLE,
+            TribeRoles.PCV_CONTROLLER,
+            TribeRoles.GOVERNOR
+        )
+    {
+        uint256 dx = IERC20(tokenSpent).balanceOf(address(this));
+        uint256 minDy = (dx * (Constants.BASIS_POINTS_GRANULARITY - maxSlippageBps)) /
+            Constants.BASIS_POINTS_GRANULARITY;
+        IERC20(tokenSpent).safeApprove(curvePool, dx);
+        ICurvePool(curvePool).exchange_underlying(i, j, dx, minDy, tokenReceivingAddress);
+        IPCVDeposit(tokenReceivingAddress).deposit();
+    }
+
+    /// @notice withdraw ERC20 from the contract
+    /// @param token address of the ERC20 to send
+    /// @param to address destination of the ERC20
+    /// @param amount quantity of ERC20 to send
+    function withdrawERC20(
+        address token,
+        address to,
+        uint256 amount
+    ) public onlyPCVController {
+        IERC20(token).safeTransfer(to, amount);
+        emit WithdrawERC20(msg.sender, token, to, amount);
+    }
+
+    /// @notice Sets the address receiving swap's outbound tokens
+    /// @param newTokenReceivingAddress the address that will receive tokens
+    function setReceivingAddress(address newTokenReceivingAddress)
+        external
+        override
+        hasAnyOfThreeRoles(TribeRoles.SWAP_ADMIN_ROLE, TribeRoles.PCV_CONTROLLER, TribeRoles.GOVERNOR)
+    {
+        _setReceivingAddress(newTokenReceivingAddress);
+    }
+
+    function _setReceivingAddress(address newTokenReceivingAddress) internal {
+        require(newTokenReceivingAddress != address(0), "CurveSwapper: zero address");
```

Comment:
I don't think there's need for `!= address(0)` checks anymore in Solidity. One of the reasons it was introduced was because Solidity didn't used to verify the length of `msg.data`, so you could accidentally call functions with no args: https://forum.openzeppelin.com/t/removing-address-0x0-checks-from-openzeppelin-contracts/2222 

Why not also check `address(1)` etc

Reply:
I agree checking is useless, it's just copy pasta for consistency because the other Swapper contract do this check & we do it in various places to make sure we don't accidentally set state variables to the burner address (but these things should be catched with integration tests before deploy)